### DISCUSSION TITLE
MemberNotNullWhen on interface members?

### DISCUSSION THREAD

**Discussion by weitzhandler** [dotnet/csharplang#8682] [General] [answered]
Hi,

I'm trying to apply `MemberNotNullWhen` attribute on an interface member but it seems to be ignored by the compiler.

```csharp
public interface IPathSegment
{
    public int? DescriptorId { get; }

    [MemberNotNullWhen(true, nameof(DescriptorId))]
    public bool IsFinal => ...
}
```

But when I do the following I still get an error when accessing `DescriptorId`:

```csharp
var segment = GetPathSegment();
if (segment.IsFinal)
{
    int descriptorId = segment.DescriptorId; // error - cannot implicitly convert int? to int
}
```

Is there a way to use these diagnostic attributes with interfaces?

**Comment by HaloFour** [comment, marked as answer]
Those attributes apply to nullable reference types, not value types.  You still have to cast from `int?` to `int` as they are two separate types.

**Reply by weitzhandler** [reply]
Is there a way around it?
Why wouldn't we instruct the compiler to use it as `int` if the value is guaranteed not to be null? Of what help is the variable if we still need to cast it to nullable?

**Reply by TahirAhmadov** [reply]
@weitzhandler the nullable value types existed before nullable reference types and the nullability state analysis. If C# was made from scratch, they could (potentially) make this scenario smoother, but at this point, it's impossible.

Minor correction to what @HaloFour wrote: the attributes apply to nullable value types, too, just differently: when you access the `.Value` property, if it thinks that it's not null, the `.Value` will not warn, whereas otherwise it would. 
Try changing it to `int descriptorId = segment.DescriptorId.Value;` and let us know if it still warns.
